modules/tensor_mechanics/src/actions/TensorMechanicsAction.C

```diff
@@ -44,87 +74,186 @@ TensorMechanicsAction::TensorMechanicsAction(const InputParameters & params) :
-
-    params.set<unsigned int>("component") = i;
-    params.set<NonlinearVariableName>("variable") = _displacements[i];
+    const std::string type = "GeneralizedPlaneStrainAction";
+    auto params = _action_factory.getValidParams(type);
+    // auto action = MooseSharedNamespace::static_pointer_cast<MooseObjectAction>(_action_factory.create("type", name() + "_gps", params));
+    //
+    // // Set the object parameters
+    // InputParameters & object_params = action->getObjectParams();
+    // object_params.set<bool>("_built_by_moose") = true;
+    //
+    // // Add the action to the warehouse
+    // _awh.addActionBlock(action);
+  }
 
-    if (_save_in.size() != 0)
-      params.set<std::vector<AuxVariableName> >("save_in") = {_save_in[i]};
-    if (_diag_save_in.size() != 0)
-      params.set<std::vector<AuxVariableName> >("diag_save_in") = {_diag_save_in[i]};
+  //
+  // Add variables (optional)
+  //
+  if (_current_task == "add_variable" && getParam<bool>("add_variables"))
+  {
+    // determine necessary order
+    const bool second = _problem->mesh().hasSecondOrderElements();
 
-    _problem->addKernel(tensor_kernel_type, kernel_name, params);
+    // Loop through the displacement variables
+    for (const auto & disp :  _displacements)
+    {
+      // Create displacement variables
+      _problem->addVariable(disp,
+                            FEType(Utility::string_to_enum<Order>(second ? "SECOND" : "FIRST"),
+                                   Utility::string_to_enum<FEFamily>("LAGRANGE")),
+                            1.0, subdomain_ids.empty() ? nullptr : &subdomain_ids);
+    }
   }
-}
 
-std::string
-TensorMechanicsAction::getKernelType()
-{
-  std::string type;
+  //
+  // Add Stess and Strain Materials (optional)
+  //
+  else if (_current_task == "add_material")
+  {
+    std::string type;
 
-  // choose kernel type based on coordinate system
-  switch (_coord_system)
+    //
+    // no plane strain
+    //
+    if (_plane_strain == PlaneStrain::None)
+    {
+      std::map<std::pair<Moose::CoordinateSystemType, Strain>, std::string> type_map = {
+        { { Moose::COORD_XYZ, Strain::Small },       "ComputeSmallStrain" },
+        { { Moose::COORD_XYZ, Strain::Incremental }, "ComputeIncrementalSmallStrain" },
+        { { Moose::COORD_XYZ, Strain::Finite },      "ComputeFiniteStrain" },
+        { { Moose::COORD_RZ, Strain::Small },        "ComputeAxisymmetricRZSmallStrain" },
+        { { Moose::COORD_RZ, Strain::Incremental },  "ComputeAxisymmetricRZIncrementalStrain" },
+        { { Moose::COORD_RZ, Strain::Finite },       "ComputeAxisymmetricRZFiniteStrain" },
+        { { Moose::COORD_RSPHERICAL, Strain::Small },        "ComputeRSphericalSmallStrain" },
+        { { Moose::COORD_RSPHERICAL, Strain::Incremental },  "ComputeRSphericalIncrementalStrain" },
+        { { Moose::COORD_RSPHERICAL, Strain::Finite },       "ComputeRSphericalFiniteStrain" }
```

Comment:
We will also need to accomodate the 1D axisymmetric case.  We don't have all of the models available yet, and the one we have needs to be renamed (currently `ComputeAxisymmetricRZIncrementalPlaneStrain`, but should be `ComputeAxisymmetric1DIncrementalSmallStrain`). Unfortunately, I think that would break this nifty code because the coord type would still be `COORD_RZ` in that scenario. The thing that differentiates it from the other axisymmetric cases would be the dimensionality of the model.